### ISSUE TITLE
open-output-file forbidden write access to ... Racket/8.1/pkgs/_LOCKpkgs.rktd

### ISSUE THREAD

**Issue by BlameTroi** [racket/racket#3851]
Windows 10, I've been using Racket for a couple of weeks and I just noticed this message. Exact text:

open-output-file: forbidden (write) access to C:\Users\Admin\AppData\Roaming\Racket\8.1\pkgs\_LOCKpkgs.rktd

I'm taking Dan Grossman's PL course and they provide us some shell files for homework: a file to edit with our code for the assignments, and a matching file containing rackunit tests. The error appears when I open the testing file. It requires rackunit and rackunit/text-ui.

I checked file and directory permissions and everything seems in order.

I installed racket 8.1 taking defaults except for the bin directory, which I did not place under the normal program files.

I'm not seeing anything interesting in windows logs. I didn't find anything via web search or in the issues that looked like a match.

I can provide the shell files if needed, but they are trivial. I am able to work with the system, but want to fix things at my end if there's a configuration problem I can address. Thanks.

**Comment by LiberalArtist** [comment]
I don't know if this is your problem, but I've seen this symptom when this assumption:

https://github.com/racket/racket/blob/bf768e10932bedbfb93afcd4f8abd1c0e4c63525/racket/collects/pkg/private/lock.rkt#L42-L51

does not hold. I was in a state where:

1. The directory existed;
2. `pkgs.rktd` did not exist;
3. The lock file did not exist; and
4. The directory was on a read-only filesystem.

I'm not sure № 2 is ever supposed to happen: I was deep in the weeds of configuring Racket installation options. But it seems the same problem would happen even if `pkgs.rktd` did exist.

**Comment by BlameTroi** [comment]
I'll check items 2 and 3 on my system. I know the directory exists (1) and the file system is read-write and I have full access to the directory tree (4). Thanks.

**Comment by BlameTroi** [comment]
A closer look shows the directories under ~/AppData/Racket were marked read only. Removing the r/o flag resolves my problem, but is this something install should handle better? I'm quite comfortable crawling around and finding hidden directories, but I'm not sure everyone who might want to use Racket is.

My issue is resolved, so I'm closing this. Thanks for the comment that motivated a deeper look @LiberalArtist.

**Comment by LiberalArtist** [comment]
> Removing the r/o flag resolves my problem, but is this something install should handle better? I'm quite comfortable crawling around and finding hidden directories, but I'm not sure everyone who might want to use Racket is.

Yes, if Racket got your installation into that state, I'd consider it a Racket bug. But I have a few more questions.

> open-output-file: forbidden (write) access to C:\Users\Admin\AppData\Roaming\Racket\8.1\pkgs_LOCKpkgs.rktd

> A closer look shows the directories under ~/AppData/Racket were marked read only.

I'm very much not a regular Windows user, but I've supported colleagues who are, and I happen to have just installed Racket on Windows a few days ago.

Is `Admin` your user name? Using "Run as administrator" in the wrong sort of pattern can definitely cause problems like this.

Could you share your output for the following program?
```racket
#lang racket

(require setup/dirs
         pkg
         pkg/lib)

(find-user-pkgs-dir)
(get-pkgs-search-dirs)
(default-pkg-scope)
(pkg-show-command #:long #t)
```
On my fresh Windows installation, I get:
```
#<path:C:\Users\philip\AppData\Roaming\Racket\8.1\pkgs>
'(#<path:C:\Program Files\Racket\share\pkgs>)
'user
Installation-wide:
 Package              Checksum                                    Source
 main-distribution    791c98e6fdf47c5845c06aca47c21606768a3402    (catalog "main-distribution")
 racket-lib           514d5381d607c78cfb2ca3ecab7deace7b890ff6    (catalog "racket-lib")
 [204 auto-installed packages not shown]
User-specific for installation "8.1":
 [none]
```

**Comment by BlameTroi** [comment]
Yes, my user name is Admin. The build house I bought it from years ago did that and I decided it wasn't worth changing. And it really is the main account on the system but I still have to do the "run as admin" thing for many operations.  I don't do so unless I'm doing something that would legitimately require doing so (installs, some powershell scripts, etc.

I would not have set the system up this way, but as it's a personal single user system I left it as is.

Here's the output you requested.

Welcome to DrRacket, version 8.1 [cs].
Language: racket, with debugging; memory limit: 128 MB.
#<path:C:\Users\Admin\AppData\Roaming\Racket\8.1\pkgs>
'(#<path:E:\bin\Racket\share\pkgs>)
'user
Installation-wide:
 Package              Checksum                                    Source
 main-distribution    791c98e6fdf47c5845c06aca47c21606768a3402    (catalog "main-distribution")
 racket-lib           514d5381d607c78cfb2ca3ecab7deace7b890ff6    (catalog "racket-lib")
 [204 auto-installed packages not shown]
User-specific for installation "8.1":
 [none]


My %USERPROFILE% directory (home, appdata sits under here) is the windows default, C:\Users\Admin\.

**Comment by VladislavNekto** [comment]
I got this error on openSuse Leap, with Racket 7.3. 
`/usr/share/racket/collects/racket/private/more-scheme.rkt:261:28: open-output-file: cannot open output file
  path: /usr/share/racket/pkgs/.LOCKpkgs.rktd
  system error: Permission denied; errno=13`
 I think, it's normal, when user can't write in /, but:
 `/usr/share/racket$ ls -la | grep pkgs
drwxr-xr-x 1 root root  5K Sep 23 19:35 pkgs`
I can easy fix it with `chmod -R /usr/share/racket/pkgs`, but it's wrong, i guess.

**Comment by capfredf** [comment]
I am on Tumble Weed. I have run into this problem too. My initial investigation shows that on Fedora and Ubuntu, the package maintainers have added `.LOCKpkgs.rktd` to the directory. I think we probably need to do the same thing on OpenSuse.

**Comment by capfredf** [comment]
@VladislavNekto for the time being, you can just create an empty file as `.LOCKpkgs.rktd` in `/usr/share/racket/pkgs`

**Comment by capfredf** [comment]
I don't know why the suse's Racket package maintainer chose to delete the file 
see Line 164 in [the build file](https://build.opensuse.org/package/view_file/openSUSE:Factory/racket/racket.spec?expand=1)

**Comment by LiberalArtist** [comment]
When packaging "minimal Racket" for Guix, I removed the `share/racket/pkgs` directory as a workaround ([lines 232–244 of gnu/packages/racket.scm](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/racket.scm?id=34cf1f451e30727b8af1de8c5e033852664e30a1#n232)):
```scheme
         (add-after 'install 'remove-pkgs-directory
           ;; If the configured pkgs-dir exists, "pkgs.rktd" does not
           ;; exist, and a lock file does not exist, commands like
           ;; `raco pkg show` will try to create a lock file and fail
           ;; due to the read-only store.
           ;; Arguably this may be a bug in `pkg/private/lock`:
           ;; see <https://github.com/racket/racket/issues/3851>.
           ;; As a workaround, remove the directory.
           (lambda* (#:key outputs #:allow-other-keys)
             ;; rmdir because we want an error if it isn't empty
             (rmdir (string-append (assoc-ref outputs "out")
                                   "/share/racket/pkgs"))
             #t)))))
```

I'm increasingly convinced that the logic in `pkgs/private/lock` I [discussed above](https://github.com/racket/racket/issues/3851#issuecomment-850595340) is flawed and should be fixed. I would reopen this issue if I had the necessary permissions.

**Comment by samth** [comment]
I've reopened this issue. Having packagers manually add a lock file or remove directories sounds really bad.

**Comment by mflatt** [comment]
@LiberalArtist After looking at this again, I agree that it would be ok to change `(directory-exists? d)` in "lock.rkt" to

```
   (and (directory-exists? d)
        (or (not read-only?)
            (file-exists? (pkg-db-file))))
```

Do I understand correctly that you're recommending that kind of change?

**Comment by LiberalArtist** [comment]
@mflatt That sounds right. I'm slightly hesitant, just because I'm not sure I've paged back in all of the various possible states this can get into, but accounting for the fact that the installation-scoped package directory may be read-only seems like at least a large part of the fix.

As far as what we were discussing on Slack, it makes sense to me for "minimal Racket" to include `base` and `racket-lib`, and *not* supporting an extra configuration also seems like a worthy goal. I was surprised to find that different ways of installing "minimal Racket" differed in this respect. One consideration I haven't fully thought through is the role of the package version in ensuring that there is a compatible package catalog.
